cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -417,35 +433,91 @@ struct GrouperImpl : public Grouper {
       RETURN_NOT_OK(encoders_[i]->Encode(batch[i], batch.length, key_buf_ptrs.data()));
     }
 
-    TypedBufferBuilder<uint32_t> group_ids_batch(ctx_->memory_pool());
-    RETURN_NOT_OK(group_ids_batch.Resize(batch.length));
+    using MapIterator = typename decltype(map_)::iterator;
 
-    for (int64_t i = 0; i < batch.length; ++i) {
-      int32_t key_length = offsets_batch[i + 1] - offsets_batch[i];
-      std::string key(
-          reinterpret_cast<const char*>(key_bytes_batch.data() + offsets_batch[i]),
-          key_length);
-
-      auto it_success = map_.emplace(key, num_groups_);
-      auto group_id = it_success.first->second;
-
-      if (it_success.second) {
-        // new key; update offsets and key_bytes
-        ++num_groups_;
-        // Skip if there are no keys
-        if (key_length > 0) {
-          auto next_key_offset = static_cast<int32_t>(key_bytes_.size());
-          key_bytes_.resize(next_key_offset + key_length);
-          offsets_.push_back(next_key_offset + key_length);
-          memcpy(key_bytes_.data() + next_key_offset, key.c_str(), key_length);
+    struct LookupResult {
+      bool inserted;
+      bool found;
+      MapIterator it;
+    };
+
+    auto generate_keys = [&](auto&& lookup_key, auto&& visit_group,
```

Comment:
Should we consider factoring these lambdas out as individual private member functions? This could avoid `ConsumeImpl` being too long.

Reply:
I'll try to, but given they also access local state, I don't know how ergonomic it will end up.

Reply:
I think I've improved this quite a bit now, can you take a look again?